### PR TITLE
Configurable Private Cluster Dns/Fqdn

### LINKED ISSUE #235: Private Cluster DNS Fqdn
Need to complete the UI for configuring control plane DNS for private clusters

### LINKED ISSUE #328: Bump default k8s version
The default kubernetes version for AKS in the Azure portal is now 1.22.11, we need to bump the default up from 1.22.6

### PR DESCRIPTION
**Pull request by Gordonby**
## PR Summary

Introduces 2 new parameters to provide more control over Private Cluster DNS
1. privateClusterDnsMethod
2. dnsApiPrivateZoneId

Closes #235 (Private Cluster Fqdn)
Closes #328 (K8S Version)

### New default behaviour for private clusters

image

mc_ resource group
image

### Behaviour for custom dns clusters

image

image

### UI Changes

image

### TODO

- [x] Discuss new parameters and logic
- [x] Build bicep code
- [x] Make UI changes
- [x] UI : Add validation of value (regex) for Dns Zone Id
- [x] UI : Adjust preset values for existing preset scenarios
- [x] Test and screengrab different deployment configurations

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] This PR is ready to merge and is not **Work in Progress**
- [x] Link to a filed issue
- [x] Screenshot of UI changes (if PR includes UI changes)

### CODE REVIEW

**Comment by khowling** [comment]
Doesn't look like the bicep is assigning the agentpool identity roles to my dnszone, here is the paramters

```
az deployment group create -g az-k8s-3xt4-rg  --template-file ./bicep/main.bicep --parameters \
	resourceName=az-k8s-3xt4 \
	agentCount=2 \
	upgradeChannel=stable \
	JustUseSystemPool=true \
	custom_vnet=true \
	CreateNetworkSecurityGroups=true \
	enable_aad=true \
	AksDisableLocalAccounts=true \
	enableAzureRBAC=true \
	adminPrincipalId=$(az ad signed-in-user show --query id --out tsv) \
	registries_sku=Basic \
	acrPushRolePrincipalId=$(az ad signed-in-user show --query id --out tsv) \
	azurepolicy=audit \
	dnsZoneId=/subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.Network/dnszones/xxx \
	azureKeyvaultSecretsProvider=true \
	createKV=true \
	kvOfficerRolePrincipalId=$(az ad signed-in-user show --query id --out tsv)
```

File: bicep/main.bicep [hunk 1]
```diff
@@ -171,21 +170,17 @@ var appGwSubnetId = ingressApplicationGateway ? (custom_vnet ? network.outputs.a
...
-var isPrivate = !empty(dnsZoneId) ? split(dnsZoneId, '/')[7] == 'privateDnsZones' : false
+var isDnsZonePrivate = !empty(dnsZoneId) ? split(dnsZoneId, '/')[7] == 'privateDnsZones' : false
 
-module dnsZone './dnsZone.bicep' = if (!empty(dnsZoneId)) {
+module dnsZone './dnsZoneRbac.bicep' = if (!empty(dnsZoneId)) {
   name: 'addDnsContributor'
-  scope: resourceGroup(dnsZoneRg)
   params: {
-    dnsZoneName: dnsZoneName
-    isPrivate: isPrivate
```
Comment (khowling) [inline bicep/main.bicep:183]: @Gordonby i think there are a bunch of issues here, you're not passing `dnsZoneId` or any parts of it to the module

File: bicep/main.bicep [hunk 2]
```diff
@@ -171,21 +170,17 @@ var appGwSubnetId = ingressApplicationGateway ? (custom_vnet ? network.outputs.a
...
 @description('The full Azure resource ID of the DNS zone to use for the AKS cluster')
 param dnsZoneId string = ''
-var dnsZoneRg = !empty(dnsZoneId) ? split(dnsZoneId, '/')[4] : ''
-var dnsZoneName = !empty(dnsZoneId) ? split(dnsZoneId, '/')[8] : ''
-var isPrivate = !empty(dnsZoneId) ? split(dnsZoneId, '/')[7] == 'privateDnsZones' : false
+var isDnsZonePrivate = !empty(dnsZoneId) ? split(dnsZoneId, '/')[7] == 'privateDnsZones' : false
 
-module dnsZone './dnsZone.bicep' = if (!empty(dnsZoneId)) {
+module dnsZone './dnsZoneRbac.bicep' = if (!empty(dnsZoneId)) {
   name: 'addDnsContributor'
```
Comment (khowling) [inline bicep/main.bicep:176]: you have this deployment name in 2 places, better to make them distinct so can track it in the portal deployment panel

**Review by khowling — CHANGES_REQUESTED** [review]

File: bicep/main.bicep (hunk 1 above)
Comment (Gordonby) [reply bicep/main.bicep:183]: Lol. Good spot.
For some reason the module allowed that param to be optional.

All fixed up.

**Comment by Gordonby** [comment]
I've addressed your points @khowling 

What was really missing here was the BYO vnet CI/CD doing a full run. It leverages a dns name for it's application. Had it run, i would have spotted these issues earlier.  I've made a change to its CI/CD trigger so that it does a full run if the `test-deploy-byoconfig` label is on the PR

Manual test of it here:
https://github.com/Azure/AKS-Construction/runs/7404444607?check_suite_focus=true
image

Can you re-run your test again please?

**Review by khowling — APPROVED** [review]

**Pull request merged**